### PR TITLE
cmd/xdpderper,derp/xdp: implement mode that drops STUN packets

### PR DESCRIPTION
**Pull request by jwhited**
This is useful during maintenance as a method for shedding home client load.

Updates tailscale/corp#20689

### CODE REVIEW

File: cmd/xdpderper/xdpderper.go [hunk 1]
```diff
@@ -57,7 +58,26 @@ func main() {
 	log.Println("XDP STUN server started")
 
 	mux := http.NewServeMux()
-	tsweb.Debugger(mux)
+	debug := tsweb.Debugger(mux)
+	debug.KVFunc("Drop STUN", func() any {
+		return server.GetDropSTUN()
+	})
+	debug.Handle("drop-stun-on", "Drop STUN packets", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (sailorfrag) [inline cmd/xdpderper/xdpderper.go:65]: A time limited drop might be nice too, but isn't needed in this PR.

File: derp/xdp/xdp.c [hunk 2]
```diff
@@ -14,6 +14,10 @@ struct config {
 	// the context of the data. cilium/ebpf uses native endian encoding for map
 	// encoding even if we use big endian types here, e.g. __be16.
 	__u16 dst_port;
+	// If drop_stun is set to a nonzero value all UDP packets destined to
+	// dst_port will be dropped. This is useful for shedding home client load
+	// during maintenance.
+	__u16 drop_stun;
```
Comment (sailorfrag) [inline derp/xdp/xdp.c:20]: Any particular reason to use __u16 vs any other size?

File: derp/xdp/xdp_default.go [hunk 3]
```diff
@@ -26,3 +26,11 @@ func (s *STUNServer) Close() error {
...
 
 func (s *STUNServer) Collect(metricCh chan<- prometheus.Metric) {}
+
+func (s *STUNServer) SetDropSTUN(v bool) error {
+	return errors.New("unimplemented on this GOOS")
+}
+
+func (s *STUNServer) GetDropSTUN() bool {
+	return false
+}
```
Comment (sailorfrag) [inline derp/xdp/xdp_default.go:36]: Maybe it would be more honest to say that it _always_ drops STUN 😄

File: derp/xdp/xdp_linux.go [hunk 4]
```diff
@@ -262,6 +266,31 @@ func (s *STUNServer) Collect(metricCh chan<- prometheus.Metric) {
...
+	dropSTUN := 0
+	if v {
+		dropSTUN = 1
+	}
+	xdpConfig := bpfConfig{
+		DstPort:  uint16(s.dstPort),
+		DropStun: uint16(dropSTUN),
+	}
+	var key uint32
+	err := s.objs.ConfigMap.Put(key, &xdpConfig)
```
Comment (sailorfrag) [inline derp/xdp/xdp_linux.go:281]: I believe there's a slight preference to make `xdpConfig` a pointer at initialization if it's only ever used as a pointer.

**Review by sailorfrag — APPROVED** [review]

File: derp/xdp/xdp.c (hunk 2 above)
Comment (jwhited) [reply derp/xdp/xdp.c:20]: __u8 results in a padding byte, and I preferred to avoid that

File: derp/xdp/xdp_default.go (hunk 3 above)
Comment (jwhited) [reply derp/xdp/xdp_default.go]: true updated in https://github.com/tailscale/tailscale/pull/12527/commits/05369ddd7b49d22ea9506826f3fbca872bb39c76

File: derp/xdp/xdp_linux.go (hunk 4 above)
Comment (jwhited) [reply derp/xdp/xdp_linux.go]: updated in https://github.com/tailscale/tailscale/pull/12527/commits/41b5d9932dfca4bb99c8c04add93cc9ae19d34be

**Pull request merged**